text/3639-diagnostic-blocking.md

```diff
@@ -0,0 +1,82 @@
+  --> $DIR/blocking-calls-in-async.rs:5:1
+   |
+5  | #[diagnostic::blocking]
+   | ^^^^^^^^^^^^^^^^^^^^^^^
+6  | fn interesting() {}
+   | ----------------
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The barest version of this feature is the addition of the `blocking` attribute to the `diagnostic` namespace, *and nothing else*, with no behavioral change. This would allow clippy and other tools to use the mark for their own analysis.
+
+The next step up is for `rustc` to incorporate a lint for direct calls. This would be trivial to implement.
+
+The final version would be to implement a call-graph analyisis lint so that transitive calls to blockign operations also produce a warning, but this is explicitly out of scope of this RFC.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There is a risk crate authors will *over*annotate their APIs with `#[diagnostic::blocking]` to get around temporary lacks in the analysis. Crate authors might *also* overannotate their APIs as a way of discouraging calls from `async` scopes, even if they are not blocking.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The use of the `#[diagnostic]` namespace has a few benefits:
+
+ - Any library with an MSRV >= 1.78 can safely add the annotation to their APIs without concerns about their users seeing warnings or errors 
+ - We don't pollute the top-level namespace with yet another marker 
+ - The namespace is explicitly allowing for `rustc` to do *nothing* with this information. These `blocking` markers are useful even with a lack of tools using them, as they act as machine parseable documentation.
+ - `clippy` (and `rustc`) can maintain a master list of paths to check for, but such lists are usually incomplete and always unwhieldy to maintain.
+ - `clippy` could add their own attribute, but this is such a fundamental signal that I believe *should* live in the language and not in external tools.
+ - We could add a `#[rustc_blocking]` attribute, but then third party crates would not be able to use these.
+
+# Prior art
+[prior-art]: #prior-art
+
+`rustc` currently has a number of `#[rustc_*]` attributes used to influence errors and lints. The language also has attributes like `#[must_use]` which are similar in spirit to the proposed attribute. `clippy` also has explicit checks for type paths of both `std` and third party crates types in order to lint against specific invalid uses.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+
+# Future possibilities
```

Comment:
Another future possibility that I think is worth mentioning: adding syntax to the attribute to (extensibly) classify the kind or degree of blocking, so as to be able to resolve “is this really a blocking operation” edge cases by writing down “it is if you call *this* blocking” in some fashion.

```
#[diagnostic::blocking(acquires_lock)]

#[diagnostic::blocking(filesystem)]

#[diagnostic::blocking(stderr)]

#[diagnostic::blocking(::rayon::block_on_task)]
```

A note on that last one: if you're using Rayon, you care a lot about the difference between “this might block on the completion of another task in the Rayon thread pool” (great when inside the pool due to being able to yield to other tasks using the Rayon-specific mechanism, but bad from an async function) and “this might block acquiring a lock” (very dangerous to do from in the pool, because it can lead to deadlocks, as well as more ordinary CPU starvation due to blocking inside a thread-per-core pool).

Much like async, if you're not already attuned to this class of issue, it can be easy to make a mistake by accident, and it's hard to audit for, so being able to use the same kind of “don't call X from Y because you could deadlock/starve” analysis but with a narrower classification would be quite useful.